chain/src/main/scala/org/bitcoins/chain/blockchain/ChainHandler.scala

```
@@ -92,24 +110,156 @@ case class ChainHandler(
+      blockHash: DoubleSha256DigestBE)(
+      implicit ec: ExecutionContext): Future[ChainApi] = {
+
+    val filterHashBE = golombFilter.hash.flip
+
+    def validateAndInsert(
+        filterHeader: CompactFilterHeaderDb,
+        filerOpt: Option[CompactFilterDb]): Future[CompactFilterDb] = {
+      if (filterHashBE != filterHeader.filterHashBE) {
+        Future.failed(new RuntimeException(
+          s"Filter hash does not match: ${filterHashBE} != ${filterHeader.filterHashBE}"))
+      } else {
+        filerOpt match {
+          case Some(filter) =>
+            if (filter.golombFilter != golombFilter)
+              logger.error(s"Filter does not match: ${golombFilter} != ${filter.golombFilter}")
+//              Future.failed(new RuntimeException(
+//                s"Filter does not match: ${golombFilter} != ${filter.golombFilter}"))
+//            else
+              Future.successful(filter)
+          case None =>
+            val filterDb = CompactFilterDbHelper.fromGolombFilter(golombFilter, filterHeader.blockHashBE, filterHeader.height)
+            for {
+              res <- filterDAO.create(filterDb)
+            } yield res
+        }
+      }
+    }
+
+    for {
+      filterHeaderOpt <- filterHeaderDAO.findByBlockHash(blockHash)
+      filterHeader = filterHeaderOpt.getOrElse(
+        throw new RuntimeException(
+          s"Cannot find a filter header for block hash ${blockHash}"))
+      filerOpt <- filterDAO.findByHash(filterHashBE)
+      _ <- validateAndInsert(filterHeader, filerOpt)
+    } yield {
+      this
+    }
+  }
+
+  override def processCheckpoint(
+      filterHeaderHash: DoubleSha256DigestBE,
+      blockHash: DoubleSha256DigestBE)(
+      implicit ec: ExecutionContext): Future[ChainApi] = {
+    Future {
+      blockFilterCheckpoints.get(blockHash) match {
+        case Some(oldFilterHeaderHash) =>
+          if (filterHeaderHash != oldFilterHeaderHash)
+            throw new RuntimeException(
```

nit: It is more efficient here to use `Future.failed` and `Future.successful` over wrapping in a `Future` because this way no exceptions need to be caught with `try`/`catch`

---

I think this also avoids the "swallowing" exceptions that Roman has talked about, unfortunately we haven't been very diligent about that in our code bases :-(